processor/logdedupprocessor/processor_test.go

```diff
@@ -221,3 +236,214 @@ func Test_unsetLogsAreExportedOnShutdown(t *testing.T) {
+		ExcludeFields: []string{
+			fmt.Sprintf("%s.remove_me", attributeField),
+		},
+	}
+
+	// Create a processor
+	p, err := newProcessor(cfg, logsSink, processortest.NewNopSettings())
+	p.conditions = getConditions(t, cfg.Conditions)
+	require.NoError(t, err)
+
+	err = p.Start(context.Background(), componenttest.NewNopHost())
+	require.NoError(t, err)
+
+	// Create plog payload
+	logRecord1 := generateTestLogRecord(t, "Body of the log1")
+	logRecord2 := generateTestLogRecord(t, "Body of the log1")
+	logRecord3 := generateTestLogRecord(t, "Body of the log2")
+	logRecord4 := generateTestLogRecord(t, "Body of the log2")
+
+	// Differ by timestamps
+	logRecord1.SetTimestamp(pcommon.NewTimestampFromTime(time.Now().Add(time.Minute)))
+	logRecord2.SetTimestamp(pcommon.NewTimestampFromTime(time.Now().Add(2 * time.Minute)))
+	logRecord3.SetTimestamp(pcommon.NewTimestampFromTime(time.Now().Add(3 * time.Minute)))
+	logRecord4.SetTimestamp(pcommon.NewTimestampFromTime(time.Now().Add(4 * time.Minute)))
+
+	// Set ID attributes to use for conditions
+	logRecord1.Attributes().PutInt("ID", 1)
+	logRecord2.Attributes().PutInt("ID", 1)
+	logRecord3.Attributes().PutInt("ID", 2)
+	logRecord4.Attributes().PutInt("ID", 2)
+
+	logs := plog.NewLogs()
+	rl := logs.ResourceLogs().AppendEmpty()
+	sl := rl.ScopeLogs().AppendEmpty()
+	logRecord1.CopyTo(sl.LogRecords().AppendEmpty())
+	logRecord3.CopyTo(sl.LogRecords().AppendEmpty())
+	logRecord2.CopyTo(sl.LogRecords().AppendEmpty())
+	logRecord4.CopyTo(sl.LogRecords().AppendEmpty())
+
+	// Consume the payload
+	err = p.ConsumeLogs(context.Background(), logs)
+	require.NoError(t, err)
+
+	// Wait for the logs to be emitted
+	require.Eventually(t, func() bool {
+		return logsSink.LogRecordCount() > 2
+	}, 3*time.Second, 200*time.Millisecond)
+
+	allSinkLogs := logsSink.AllLogs()
+	require.Len(t, allSinkLogs, 2)
+
+	consumedLogs := allSinkLogs[0]
+	require.Equal(t, 2, consumedLogs.LogRecordCount())
+
+	require.Equal(t, 1, consumedLogs.ResourceLogs().Len())
+	consumedRl := consumedLogs.ResourceLogs().At(0)
+	require.Equal(t, 1, consumedRl.ScopeLogs().Len())
+	consumedSl := consumedRl.ScopeLogs().At(0)
+	require.Equal(t, 2, consumedSl.LogRecords().Len())
+
+	for i := 0; i < consumedSl.LogRecords().Len(); i++ {
+		consumedLogRecord := consumedSl.LogRecords().At(i)
+		ID, ok := consumedLogRecord.Attributes().Get("ID")
+		require.True(t, ok)
+		require.Equal(t, int64(2), ID.Int())
+	}
+
+	dedupedLogs := allSinkLogs[1]
+	require.Equal(t, 1, dedupedLogs.LogRecordCount())
+
+	require.Equal(t, 1, dedupedLogs.ResourceLogs().Len())
+	dedupedRl := dedupedLogs.ResourceLogs().At(0)
+	require.Equal(t, 1, dedupedRl.ScopeLogs().Len())
+	dedupedSl := dedupedRl.ScopeLogs().At(0)
+	require.Equal(t, 1, dedupedSl.LogRecords().Len())
+	dedupedLogRecord := dedupedSl.LogRecords().At(0)
+
+	countVal, ok := dedupedLogRecord.Attributes().Get(cfg.LogCountAttribute)
+	require.True(t, ok)
+	require.Equal(t, int64(2), countVal.Int())
```

Comment:
When comparing actual vs expected telemetry, the code can be very complex and still the comparison can be very incomplete. We added some packages to facilitate more precise validation of expected results.

Specifically, use the `golden` package capture and reload expected results. Use `golden.WriteLogs` to write these objects once. Then delete the `WriteLogs` call and use `ReadLogs` in the actual test.

Then use `pdatatest/plogtest.CompareLogs` (and make sure to assert no error). You may need to pass in some options to ignore volatile parts of the logs.

[Here's an example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7f65b87fcde5e9f61088a9903ee35433bcdaa0eb/connector/otlpjsonconnector/connector_test.go#L59).

Reply:
Good suggestion, just pushed a commit to use golden for the tests and it looks much cleaner now.